codalab/lib/upload_manager.py

```diff
@@ -0,0 +1,187 @@
+                                source_output_path, zip_util.strip_archive_ext(source_output_path),
+                                remove_source=True, simplify_archive=simplify_archives)
+                elif is_local_path:
+                    source_path = path_util.normalize(source)
+                    path_util.check_isvalid(source_path, 'upload')
+                    
+                    if unpack and self._can_unpack_file(source_path):
+                        self._unpack_file(
+                            source_path, zip_util.strip_archive_ext(source_output_path),
+                            remove_source=remove_sources, simplify_archive=simplify_archives)
+                    elif remove_sources:
+                        path_util.rename(source_path, source_output_path)
+                    else:
+                        path_util.copy(source_path, source_output_path, follow_symlinks=follow_symlinks, exclude_patterns=exclude_patterns)
+                elif is_fileobj:
+                    if unpack and zip_util.path_is_archive(filename):
+                        self._unpack_fileobj(
+                            source[0], source[1],
+                            zip_util.strip_archive_ext(source_output_path),
+                            simplify_archive=simplify_archives)
+                    else:
+                        with open(source_output_path, 'wb') as out:
+                            shutil.copyfileobj(source[1], out)
+
+            if len(sources) == 1:
+                self._simplify_directory(bundle_path)
+        except:
+            if os.path.exists(bundle_path):
+                path_util.remove(bundle_path)
+            raise
+    
+    def _interpret_source(self, source):
+        is_url, is_local_path, is_fileobj = False, False, False
+        if isinstance(source, basestring):
+            if path_util.path_is_url(source):
+                is_url = True
+            else:
+                is_local_path = True
+            filename = os.path.basename(os.path.normpath(source))
+        else:
+            is_fileobj = True
+            filename = source[0]
+        return is_url, is_local_path, is_fileobj, filename
+
+    def _can_unpack_file(self, path):
+        return os.path.isfile(path) and zip_util.path_is_archive(path)
+    
+    def _unpack_file(self, source_path, dest_path, remove_source, simplify_archive):
+        zip_util.unpack(zip_util.get_archive_ext(source_path), source_path, dest_path)
+        if remove_source:
+            path_util.remove(source_path)
+        if simplify_archive:
+            self._simplify_archive(dest_path)
+
+    def _unpack_fileobj(self, source_filename, source_fileobj, dest_path, simplify_archive):
+        zip_util.unpack(zip_util.get_archive_ext(source_filename), source_fileobj, dest_path)
+        if simplify_archive:
+            self._simplify_archive(dest_path)
+
+    def _simplify_archive(self, path):
+        if not os.path.isdir(path):
```

Comment:
Add comment 'Modify |path| in place: if |path| is a directory containing exactly one file/directory that's not ignored, then replace path with that file/directory.'